src/herad.cpp

```diff
@@ -148,7 +153,10 @@ uint16_t HSQ_decompress(uint8_t * data, int size, uint8_t * out)
 		// get next bit of the queue
 		if (queue == 1)
 		{
-			queue = *(uint16_t *)src | 0x10000;
+			// need to perform unaligned access
+			uint16_t temp;
+			memcpy(&temp, src, sizeof(uint16_t));
+			queue = temp | 0x10000;
```

Comment:
While I'm not opposed to these changes in principle, they do make the code a little more verbose.  Is there any way you could refactor things so the `memcpy` is not necessary, e.g. by making `src` a `uint16_t` array from the start, so it can be accessed directly as `*src`, and `src++` instead of `src += 2`?

Failing that, I wonder whether a macro or inline function could be used to hide the temporary variable and make the code a bit clearer?

Reply:
The problem is: Do you have any guarantees that the incoming stream will always be aligned? Say for instance we have the following datastream:

```
|-------- 4 bytes --------|- 1 bytes -|---- 2 bytes ----|
```
then the last two bytes will not be aligned correctly anymore, regardless of the fact that the initial array might have been 4-byte aligned. My suggestion is to create littler getters/setters like
```
uint16_t get_uint16_t(const unsigned char* src, endianness = LITTLE)
{
  uint16_t res;
  memcpy(&res, src, sizeof(res));
  return res;
}

void set_uint16_t(uint16_t& dest, const unsigned char* src, endianness = LITTLE)
{
  dest = get_uint16_t(src, endianness);
}
```
which allows cleaning up the whole mess in one go, and memcpy is nicely tucked away in one function. What do you think?

Reply:
You are correct that if the incoming stream has different bit-widths then you will need something like this.

However all your proposed changes in `herad.cpp` are `uint16_t` which makes me wonder whether you could just change `src` to be a `uint16_t` array, and load it element by element with `readInt(2)`.  This way the platform may allocate it as an array of 32-bit fields in order to allow direct element access, but this won't matter as you're never accessing the memory directly if you treat it as a normal array.  It will also fix endian issues as the bytes will be loaded in the correct order from the start.  For that matter you could forget the size-specific component and simply allocate it as an array of `unsigned int`.  However I haven't looked at the code in detail so not sure whether this will work.

This does beg the question that if C/C++ cannot do unaligned memory accesses, what actually happens if you have an array of `uint16_t`?  Is it padded in memory or does the compiler generate extra instructions to allow unaligned accesses?

Reply:
@Malvineous checkout line 160/168:
```
// copy next byte of the input to the output
*dst++ = *src++;
```
so clearly `src` points to a stream of intermingled `uint8_t`s and `uint16_t`s, which means all alignment guarantees will be out the window at *some point* during this parse loop.

**Valid** C/C++ code will never do unaligned access, because the compiler will insert padding wherever necessary (which is why you should avoid packed structs like the plague). For instance, a normal `uint16_t arr[N]` will **always** be 2 byte aligned on literally every platform but the msot crude 8-bit CPUs. The problem here is that you have a stream of lots of different things, and at some point you're treating a bunch of 4 bytes like an `uint32_t` and doing pointer casts, which violates 1) strict aliasing in C/C++ 2) more often than not violates the alignment requirements on C. The extreme example here is SPARC: I tried, this code crashes on SPARC due to a `SIGBUS`, as SPARC *does not support unaligned access*.

Reply:
Yes I can see my thought of changing `src` to be a `uint16_t` array is probably more trouble than it's worth, if the compiler optimises away the `memcpy` call.

Your proposed `get_uint16_t` function looks good, although declaring it `inline` might help optimisations.  Perhaps we could call the function something like `u16_unaligned()` just as a reminder why it's there?  I'm the sort of programmer who is likely to replace the function with typecasts one day thinking they are more efficient, and completely forgetting about the alignment issues.  This will also avoid the need to comment each use.

What do you think?

Reply:
Sure, I while work on that then. I can only work on it on the weekend, and will change it as discussed.
(sidenote: `inline` has negligible effect on whether the function actually gets inlined. `inline` is important in C++ for reasons of ODR (multiple, but equal, definitions of the same function), whereas in C the semantics are different but still do not really affect inlining but whether an eternal symbol is emitted at all).